iwyu_include_picker.cc

```diff
@@ -1244,16 +1245,42 @@ IncludePicker::IncludePicker(RegexDialect regex_dialect,
 
 void IncludePicker::AddDefaultMappings(CStdLib cstdlib,
                                        CXXStdLib cxxstdlib) {
+  using clang::tooling::stdlib::Header;
+  using clang::tooling::stdlib::Lang;
+  using clang::tooling::stdlib::Symbol;
+
   if (cstdlib == CStdLib::Glibc) {
     AddSymbolMappings(libc_symbol_map, IWYU_ARRAYSIZE(libc_symbol_map));
     AddIncludeMappings(libc_include_map, IWYU_ARRAYSIZE(libc_include_map));
+  } else if (cstdlib == CStdLib::Canonical) {
+    // Get canonical C standard library mappings from clang tooling
```

Comment:
I absolutely love this, compared to all the matchsticks, wire and old band-aids we do for standard library include mappings. Is `precomputed_tpl_args.cc` the only failure if you hack it on to every test invocation?

I wonder if we should look into fixing that first, and then just making this the default and only mapping set for the standard library.

Reply:
With MacOS (and libcxx), I had the canonical mappings on by default for a run, and the only extra test failure was `precomputed_tpl_args.cc` (there were still 5 other failures that I was already seeing). Not sure if that's the case with libstdc++.

Running on normal code though, you also get a few things leaking from private libcxx headers - `operator<` for some type is one I recall seeing. I expect we need to check what (if anything) is being reported for that via cppreference, and if IWYU wants it in a slightly different style to 'match' properly.

The existing mappings mainly rely on file-mappings - this is nice because the full function prototypes are matched. I wouldn't be surprised if there's a reasonable amount of fallout from switching to mainly symbol-spelling matches.

As I mentioned previously, I'd be inclined to merge more or less as is (happy to put this under an exprimental arg) - so everyone can play with it, and see what problems this has in the wild.

Reply:
urk, I misremembered. There are a few other failures in the tests when forcing the canonical mappings, at least badinc, check_also, no_char_traits also fail. Some of these also fail on macos anyway, but I think the output may have changed too.

Reply:
Here are the new failures on my Ubuntu system (normally all tests pass):
```
======================================================================
FAIL: test_badinc (__main__.cxx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 88, in <lambda>
    setattr(cls, test_name, lambda x, f=filename: TestIwyuOnRelevantFiles(f))
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 64, in TestIwyuOnRelevantFiles
    iwyu_test_util.TestIwyuOnRelativeFile(filename, files_to_check, verbose=True)
  File "/home/kimgr/code/iwyu/include-what-you-use/iwyu_test_util.py", line 478, in TestIwyuOnRelativeFile
    raise AssertionError(''.join(failures))
AssertionError: 
tests/cxx/badinc.cc:1069: Unexpected diagnostic:
va_list is defined in <cstdarg>, which isn't directly #included.

tests/cxx/badinc.cc:1069: Unmatched regex:
va_list is...*<stdio.h>

Unexpected summary diffs for tests/cxx/badinc.cc:
+++ 
@@ -1,6 +1,7 @@
 tests/cxx/badinc.cc should add these lines:
 #include <ctype.h>
 #include <stddef.h>
+#include <cstdarg>
 #include <list>
 #include "tests/cxx/badinc-i1.h"
 class D2_Class;
@@ -31,7 +32,8 @@
 #include <setjmp.h>
 #include <stddef.h>  // for offsetof
 #include <algorithm>  // for find
-#include <fstream>  // for fstream
+#include <cstdarg>  // for va_list
+#include <fstream>  // for basic_fstream, fstream
 #include <list>  // for list
 #include <string>  // for basic_string, operator+, string
 #include <typeinfo>  // for type_info
---


======================================================================
FAIL: test_check_also (__main__.cxx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 88, in <lambda>
    setattr(cls, test_name, lambda x, f=filename: TestIwyuOnRelevantFiles(f))
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 64, in TestIwyuOnRelevantFiles
    iwyu_test_util.TestIwyuOnRelativeFile(filename, files_to_check, verbose=True)
  File "/home/kimgr/code/iwyu/include-what-you-use/iwyu_test_util.py", line 478, in TestIwyuOnRelativeFile
    raise AssertionError(''.join(failures))
AssertionError: 
tests/cxx/check_also-d1.h:18: Unexpected diagnostic:
NULL is defined in <cstddef>, which isn't directly #included.

tests/cxx/check_also-d1.h:18: Unmatched regex:
NULL is...*<stddef.h>

Unexpected summary diffs for tests/cxx/check_also-d1.h:
+++ 
@@ -1,8 +1,8 @@
 tests/cxx/check_also-d1.h should add these lines:
-#include <stddef.h>
+#include <cstddef>
 
 tests/cxx/check_also-d1.h should remove these lines:
 - #include "check_also-i1.h"  // lines XX-XX
 
 The full include-list for tests/cxx/check_also-d1.h:
-#include <stddef.h>  // for NULL
+#include <cstddef>  // for NULL
---


======================================================================
FAIL: test_no_char_traits (__main__.cxx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 88, in <lambda>
    setattr(cls, test_name, lambda x, f=filename: TestIwyuOnRelevantFiles(f))
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 64, in TestIwyuOnRelevantFiles
    iwyu_test_util.TestIwyuOnRelativeFile(filename, files_to_check, verbose=True)
  File "/home/kimgr/code/iwyu/include-what-you-use/iwyu_test_util.py", line 478, in TestIwyuOnRelativeFile
    raise AssertionError(''.join(failures))
AssertionError: 
tests/cxx/no_char_traits.cc:29: Unexpected diagnostic:
std::operator<< is defined in <ostream>, which isn't directly #included.

tests/cxx/no_char_traits.cc:31: Unexpected diagnostic:
std::char_traits is defined in <ostream>, which isn't directly #included.

tests/cxx/no_char_traits.cc:31: Unexpected diagnostic:
std::operator<< is defined in <ostream>, which isn't directly #included.

Unexpected summary diffs for tests/cxx/no_char_traits.cc:
+++ 
@@ -1 +1,8 @@
-(tests/cxx/no_char_traits.cc has correct #includes/fwd-decls)
+tests/cxx/no_char_traits.cc should add these lines:
+#include <ostream>
+
+tests/cxx/no_char_traits.cc should remove these lines:
+
+The full include-list for tests/cxx/no_char_traits.cc:
+#include <iostream>  // for basic_ostream, cout
+#include <ostream>  // for char_traits, operator<<
---


======================================================================
FAIL: test_precomputed_tpl_args (__main__.cxx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 88, in <lambda>
    setattr(cls, test_name, lambda x, f=filename: TestIwyuOnRelevantFiles(f))
  File "/home/kimgr/code/iwyu/include-what-you-use/./run_iwyu_tests.py", line 64, in TestIwyuOnRelevantFiles
    iwyu_test_util.TestIwyuOnRelativeFile(filename, files_to_check, verbose=True)
  File "/home/kimgr/code/iwyu/include-what-you-use/iwyu_test_util.py", line 478, in TestIwyuOnRelativeFile
    raise AssertionError(''.join(failures))
AssertionError: 
tests/cxx/precomputed_tpl_args.cc:43: Unexpected diagnostic:
std::less is defined in <functional>, which isn't directly #included.

tests/cxx/precomputed_tpl_args.cc:43: Unmatched regex:
std::less is...*precomputed_tpl_args-i1.h

tests/cxx/precomputed_tpl_args.cc:47: Unexpected diagnostic:
std::less is defined in <functional>, which isn't directly #included.

Unexpected summary diffs for tests/cxx/precomputed_tpl_args.cc:
+++ 
@@ -1,12 +1,14 @@
 tests/cxx/precomputed_tpl_args.cc should add these lines:
+#include <functional>
 #include "tests/cxx/precomputed_tpl_args-i1.h"
 
 tests/cxx/precomputed_tpl_args.cc should remove these lines:
 
 The full include-list for tests/cxx/precomputed_tpl_args.cc:
 #include <bitset>  // for bitset
+#include <functional>  // for less
 #include <map>  // for map
 #include <set>  // for set
 #include <vector>  // for vector
-#include "tests/cxx/precomputed_tpl_args-d1.h"  // for D1SpecializationClass, less
-#include "tests/cxx/precomputed_tpl_args-i1.h"  // for IndirectClass, SpecializationClass, less
+#include "tests/cxx/precomputed_tpl_args-d1.h"  // for D1SpecializationClass
+#include "tests/cxx/precomputed_tpl_args-i1.h"  // for IndirectClass, SpecializationClass
---


----------------------------------------------------------------------
Ran 178 tests in 11.693s

FAILED (failures=4)
```

Reply:
I also see a regression in test runtime, from ~10.5 to ~11.5 seconds, so around 10%. Nothing here stood out as particularly slow, so I wonder where that time goes...

Reply:
Those errors look similar, but not exaclty the same, as those that I'm seeing on MacOS.

> I also see a regression in test runtime

There are about 5200 symbols being mapped (934 C, 4221 C++). There might be time setting those up compared to the 100's we usually create? I hacked out the setup loop - for me that dropped runtime from 2.39s to 2.32s (for just processing badinc), so only accounting for about half of what you're seeing.

Reply:
Hmm. Maybe time to look into some mapping optimizations after this has landed. I don't think the regression is bad enough to do anything up-front.

Reply:
There's actually a fun mix of problems there:

* `badinc.cc` has `va_list is defined in <cstdarg>, which isn't directly #included.` -- `va_list` is only mapped to `<cstdarg>` in the clang mappings, but we accept it from a few other headers, `<stdio.h>` most notably.
* `check_also-d1.h` wants `NULL` from `<stddef.h>`, but now `<cstddef>` is suggested (for some reason, not sure why; can't find any Clang mappings for `NULL`...?)
* `no_char_traits.cc` wants `<ostream>` for `operator<<` and `std::char_traits`. I think both of these were masked by public-to-public mappings from `<ostream>` to `<iostream>` on master
* `precomputed_tpl_args.cc:43` complains about `std::less` now. I believe this used to work because `std::less` was found in some private `<bits/...map...>` header, which was eventually mapped back to `<map>`. But not sure.

I don't see any obvious systematic fixes there... Some could probably be accepted as the "right" behavior, and fixed up in the tests, but many are simply a matter of practicality.

Reply:
FWIW, macos+libcxx behaves similarly for the first 3 tests. So I suspect a few things are in the GCC mappings mappings that we don't have on libcxx / clang. I wanted to double check what each test was doing before proposing anything.

Reply:
Thanks! I think the patch looks great in shape and form. I wonder how we should think about moving forward with it. Do you think it's reasonable to augment Clang's symbol mappings with manual overlays to preserve IWYU behavior? Or should we just be less pragmatic in favor of simplicity? All of the examples brought to light by the test suite are pretty nice with the mainline built-in mappings, and pretty annoying with the Clang mappings.

Reply:
In general I expect we should follow what the mainline behavior is doing, except where we can identify the new behavior as preferable. I'm hoping we can do that with mappings. Possibly, if we inspect the symbol namings that clang is outputting, there might be a few things we can massage to work better for IWYU (or change upstream)